raft/raft.go

```diff
@@ -611,6 +611,11 @@ func (r *raft) becomePreCandidate() {
 	// but doesn't change anything else. In particular it does not increase
 	// r.Term or change r.Vote.
 	r.step = stepCandidate
+	if r.checkQuorum {
```

Comment:
I don't think this should be based on `r.checkQuorum`. That implies that we are doing this because we know that the leader is failing its quorum check, but we can't know that. The leader could be in the majority side of the partition and still be acting as leader.

#9586 discussed setting `r.lead = None` in `becomePreCandidate` whether `checkQuorum` is true or not. I think that could work here by keeping things from getting permanently stuck, but it has the odd effect that the [second (pre-)campaigner wins](https://github.com/etcd-io/etcd/issues/9586#issuecomment-383111567). For that reason I still think it would be better to modify the definition of `inLease` at line 711 by adding `&& r.state == StateFollower`.

On the other hand, that's a more complex and invasive change, and I think this version is simple and safe. It makes sense that pre-candidates act like they don't know the leader (same as candidates). I was previously concerned that it would be unusual for a node to transition between `r.lead` set and unset within a single term, but leaders already do that for `checkQuorum` so it should be fine.